src/api_manager/context/request_context.cc

```diff
@@ -292,6 +301,49 @@ void RequestContext::FillHttpHeaders(const Response *response,
   }
 }
 
+void RequestContext::FillJwtPayloads(service_control::ReportRequestInfo *info) {
+  auto serverConfig = service_context_->config()->server_config();
+  if (serverConfig->has_service_control_config() &&
+      serverConfig->service_control_config().log_jwt_payload().size() != 0) {
+    JwtCache &jwt_cache = service_context()->jwt_cache();
+    JwtCache::ScopedLookup lookup(&jwt_cache, auth_token_);
+    // FillJwtPayloads is called when reporting to service control, which is
+    // after CheckAuth, so the userInfo should always be cached.
+    if (!lookup.Found()) {
+      return;
+    }
+    const auto &jwt_payloads =
+        serverConfig->service_control_config().log_jwt_payload();
+
+    for (const auto &payload : jwt_payloads) {
+      char *json_copy = strdup(lookup.value()->user_info.claims.c_str());
+      grpc_json *property_json =
+          grpc_json_parse_string_with_len(json_copy, strlen(json_copy));
+
+      std::vector<std::string> keys;
+      split(payload, kJwtPayloadsDelimeter, &keys);
+      for (const auto &key : keys) {
+        const grpc_json *next = GetProperty(property_json, key.c_str());
+        if (next) {
+          *property_json = *next;
+        } else {
+          // Not found the corresponding jwt payload.
+          property_json = nullptr;
+          break;
+        }
+      }
+
+      if (property_json) {
+        std::string s;
+        s += property_json->key;
+        s += "=";
+        s += property_json->value;
```

Comment:
is this value always string type?

Reply:
This is a good point. 

if it is an object or a list or a bool, then property_json->value is null. Otherwise, it is char*. 

So, not good to use this grpc_json here. 

What I want here is just get the string value of a field, given the json string. 

I am thinking to use:  https://github.com/open-source-parsers/jsoncpp

Reply:
security is important.  Here the security means not crash from any json input.  
I believe protobuf code is more secure the any other json parser.

That is why: https://github.com/google/jwt_verify_lib
switch to use protobuf.

Reply:
BTW, is it possible to extract header and jwt payload logging code into a separate file so it can be easily re-used by others.

Reply:
Actually ESP has this JsonToProto method, using protobuf.Struct (https://github.com/cloudendpoints/esp/blob/master/src/api_manager/utils/marshalling.cc)

however, it is similar to grpc_json, still needs to handle seperately according to type.

So, still using grpc_json, but only supports primitive fields. 

related test code can be found: https://github.com/cloudendpoints/esp/blob/master/src/api_manager/auth/lib/json_util_test.cc

Reply:
Will follow up to refactor related code.